kpatch-build/create-diff-object.c

```diff
@@ -771,6 +775,86 @@ static bool kpatch_line_macro_change_only(struct section *sec)
 
 	return true;
 }
+
+#elif __s390x__
+
+static int insn_get_length(unsigned char *inst)
+{
+	switch (inst[0] >> 6) {
+		case 0b00:
+			return 2;
+		case 0b01:
+		case 0b10:
+			return 4;
+		case 0b11:
+		default:
+			return 6;
+	}
+}
+
+static bool kpatch_line_macro_change_only(struct section *sec)
+{
+	unsigned char *start1, *start2;
+	unsigned long size, offset, length;
+	struct rela *rela;
+	int lineonly = 0, found;
+
+	if (sec->status != CHANGED ||
+	    is_rela_section(sec) ||
+	    !is_text_section(sec) ||
+	    sec->sh.sh_size != sec->twin->sh.sh_size ||
+	    !sec->rela ||
+	    sec->rela->status != SAME)
+		return false;
+
+	start1 = sec->twin->data->d_buf;
+	start2 = sec->data->d_buf;
+	size = sec->sh.sh_size;
+	for (offset = 0; offset < size; offset += length) {
+		length = insn_get_length(start1 + offset);
+
+		if (insn_get_length(start1 + offset) != insn_get_length(start2 + offset))
+			return false;
+
+		if (!memcmp(start1 + offset, start2 + offset, length))
+			continue;
+
+		/* Verify lghi %r4 <line number> */
```

Comment:
When the line number is greater than 2 bytes,  verification of lgfi %r4 <line number> would be essential

Reply:
Hi @sumanthkorikkar : after investigating the above concern, @joe-lawrence and I found that we don't need to verify lgfi %4 when the line number is greater than 2 bytes as no c file in the kernel has lines exceeding 50k.

Verification method: find . -name '*.c' -exec wc -l {} \; | sort -n | tail -n5

Reply:
ok, fine with it. We could add that in the future, if there is a need.

Reply:
For posterity:  large numbers might result in other instruction combinations in addition to lgfi, for example from [Compiler Explorer's](https://godbolt.org/z/xv3vTnqx9) gcc 11.2.0:
```C
printf("%l\n", 0x7fff);      // lghi    %r3,32767
printf("%l\n", 0x8000);      // llill   %r3,32768
printf("%l\n", 0x10000);     // llilh   %r3,1
printf("%l\n", 0x10001);     // larl    %r13,.L3
                             // lg      %r3,.L4-.L3(%r13)
                             // .L3:
                             // .L4:
                             // .quad   65537

```